File: backend/src/models/Laudo.js
'use strict';
const { Model } = require('sequelize');

module.exports = (sequelize, DataTypes) => {
  class Laudo extends Model {
    static associate(models) {
      // Associação com User
      Laudo.belongsTo(models.User, {
        foreignKey: 'user_id',
        as: 'user'
      });
    }

    // Método para calcular status baseado no vencimento
    getStatus() {
      const hoje = new Date();
      const vencimento = new Date(this.vencimento);
      const diasParaVencimento = Math.ceil((vencimento - hoje) / (1000 * 60 * 60 * 24));
      
      if (diasParaVencimento < 0) {
        return 'vencido';
      } else if (diasParaVencimento <= 30) {
        return 'vencendo';
      } else {
        return 'vigente';
      }
    }

    // Método para calcular dias para vencimento
    getDiasParaVencimento() {
      const hoje = new Date();
      const vencimento = new Date(this.vencimento);
      return Math.ceil((vencimento - hoje) / (1000 * 60 * 60 * 24));
    }

    // Método para formatar valores monetários
    getValorSolicitadoFormatado() {
      return new Intl.NumberFormat('pt-BR', {
        style: 'currency',
        currency: 'BRL'
      }).format(this.valor_solicitado);
    }

    getValorLiberadoFormatado() {
      if (!this.valor_liberado) return null;
      return new Intl.NumberFormat('pt-BR', {
        style: 'currency',
        currency: 'BRL'
      }).format(this.valor_liberado);
    }
  }

  Laudo.init({
    id: {
      type: DataTypes.INTEGER,
      primaryKey: true,
      autoIncrement: true
    },
    parceiro: {
      type: DataTypes.STRING(255),
      allowNull: false,
      validate: {
        notEmpty: {
          msg: 'Parceiro é obrigatório'
        },
        len: {
          args: [2, 255],
          msg: 'Parceiro deve ter entre 2 e 255 caracteres'
        }
      }
    },
    tipo_imovel: {
      type: DataTypes.ENUM('casa', 'apartamento'),
      allowNull: false,
      validate: {
        isIn: {
          args: [['casa', 'apartamento']],
          msg: 'Tipo de imóvel deve ser "casa" ou "apartamento"'
        }
      }
    },
    valor_solicitado: {
      type: DataTypes.DECIMAL(15, 2),
      allowNull: false,
      validate: {
        isDecimal: {
          msg: 'Valor solicitado deve ser um número decimal'
        },
        min: {
          args: [0.01],
          msg: 'Valor solicitado deve ser maior que zero'
        }
      }
    },
    valor_liberado: {
      type: DataTypes.DECIMAL(15, 2),
      allowNull: true,
      validate: {
        isDecimal: {
          msg: 'Valor liberado deve ser um número decimal'
        },
        min: {
          args: [0],
          msg: 'Valor liberado deve ser positivo'
        }
      }
    },
    vencimento: {
      type: DataTypes.DATE,
      allowNull: false,
      validate: {
        isDate: {
          msg: 'Vencimento deve ser uma data válida'
        }
      }
    },
    endereco: {
      type: DataTypes.TEXT,
      allowNull: false,
      validate: {
        notEmpty: {
          msg: 'Endereço é obrigatório'
        },
        len: {
          args: [5, 1000],
          msg: 'Endereço deve ter entre 5 e 1000 caracteres'
        }
      }
    },
    observacoes: {
      type: DataTypes.TEXT,
      allowNull: true,
      validate: {
        len: {
          args: [0, 2000],
          msg: 'Observações não podem ter mais de 2000 caracteres'
        }
      }
    },
    arquivos: {
      type: DataTypes.JSONB,
      allowNull: true,
      defaultValue: null
    },
    user_id: {
      type: DataTypes.INTEGER,
      allowNull: false,
      references: {
        model: 'users', // ✅ CORRIGIDO: 'users' minúsculo
        key: 'id'
      }
    }
  }, {
    sequelize,
    modelName: 'Laudo',
    tableName: 'laudos', // ✅ CORRIGIDO: 'laudos' minúsculo
    timestamps: true,
    createdAt: 'created_at',
    updatedAt: 'updated_at',
    indexes: [
      {
        fields: ['parceiro']
      },
      {
        fields: ['tipo_imovel']
      },
      {
        fields: ['vencimento']
      },
      {
        fields: ['user_id']
      },
      {
        fields: ['created_at']
      }
    ]
  });

  return Laudo;
};

File: backend-go/internal/models/laudo.go
package models

import (
	"time"

	"gorm.io/datatypes"
)

// Laudo espelha a tabela `laudos` — avaliação/parecer de imóvel para um parceiro
// (banco/correspondente), com validade e valores solicitado/liberado.
// Ver docs/migration/06-dashboards-vendas-config.md §"Laudos".
//
// GOTCHA (§8): o model Node NÃO tinha `tenant_id` (escopo global + user_id).
// Corrigido deliberadamente aqui: `tenant_id` é obrigatório e filtrado sempre
// via db.WithContext(ctx) (callbacks GORM de internal/tenant).
type Laudo struct {
	ID uint `gorm:"primaryKey" json:"id"`

	Parceiro    string  `gorm:"column:parceiro;size:255;not null;index" json:"parceiro"`
	TipoImovel  string  `gorm:"column:tipo_imovel;not null;index" json:"tipo_imovel"` // casa|apartamento

	ValorSolicitado float64  `gorm:"column:valor_solicitado;not null" json:"valor_solicitado"` // DECIMAL(15,2) >0
	ValorLiberado   *float64 `gorm:"column:valor_liberado" json:"valor_liberado,omitempty"`     // DECIMAL(15,2) >=0

	Vencimento time.Time `gorm:"column:vencimento;not null;index" json:"vencimento"`
	Endereco   string    `gorm:"column:endereco;not null" json:"endereco"`
	Observacoes *string  `gorm:"column:observacoes" json:"observacoes,omitempty"`

	// Arquivos: JSONB { categoria: [{filename, originalname, path, size, mimetype}] }
	Arquivos datatypes.JSON `gorm:"column:arquivos" json:"arquivos,omitempty"`

	UserID   uint  `gorm:"column:user_id;not null;index" json:"user_id"`
	TenantID uint  `gorm:"column:tenant_id;not null;index" json:"tenant_id"` // novo — corrige gotcha §8

	CreatedAt time.Time `gorm:"column:created_at;index" json:"created_at"`
	UpdatedAt time.Time `gorm:"column:updated_at" json:"updated_at"`

	User *User `gorm:"foreignKey:UserID;references:ID" json:"user,omitempty"`
}

func (Laudo) TableName() string { return "laudos" }

// Tipos de imóvel aceitos pelo laudo.
const (
	LaudoTipoImovelCasa        = "casa"
	LaudoTipoImovelApartamento = "apartamento"
)

// IsTipoImovelValido confere se o tipo pertence ao enum casa/apartamento.
func IsTipoImovelValido(tipo string) bool {
	return tipo == LaudoTipoImovelCasa || tipo == LaudoTipoImovelApartamento
}

// Status calculados de vencimento (não persistidos — derivados em runtime).
const (
	LaudoStatusVencido  = "vencido"
	LaudoStatusVencendo = "vencendo"
	LaudoStatusVigente  = "vigente"
)

// DiasParaVencimento devolve a diferença em dias inteiros entre `vencimento` e
// `agora` (negativo = já venceu). Porta `getDiasParaVencimento()` do model Node.
func (l Laudo) DiasParaVencimento(agora time.Time) int {
	d := l.Vencimento.Truncate(24 * time.Hour).Sub(agora.Truncate(24 * time.Hour))
	return int(d.Hours() / 24)
}

// Status devolve vencido/vencendo/vigente conforme os dias restantes até o
// vencimento. Porta `getStatus()` do model Node: vencido (dias<0), vencendo
// (0<=dias<=30), vigente (dias>30).
func (l Laudo) Status(agora time.Time) string {
	dias := l.DiasParaVencimento(agora)
	switch {
	case dias < 0:
		return LaudoStatusVencido
	case dias <= 30:
		return LaudoStatusVencendo
	default:
		return LaudoStatusVigente
	}
}

// LaudoArquivoMeta é um item da lista de arquivos por categoria em `arquivos` (JSONB).
type LaudoArquivoMeta struct {
	Filename     string `json:"filename"`
	OriginalName string `json:"originalname"`
	Path         string `json:"path"`
	Size         int64  `json:"size"`
	MimeType     string `json:"mimetype"`
}
